java-checks/src/main/java/org/sonar/java/checks/security/AndroidMobileDatabaseEncryptionKeysCheck.java

```diff
@@ -0,0 +1,121 @@
+
+  private static final String JAVA_LANG_STRING = "java.lang.String";
+
+  private static final MethodMatchers SQLITE_DATABASE_CONSTRUCTOR = MethodMatchers.create()
+    .ofTypes("net.sqlcipher.database.SQLiteDatabase")
+    .constructor()
+    .withAnyParameters()
+    .build();
+
+  private static final MethodMatchers SQLITE_DATABASE_METHODS = MethodMatchers.create()
+    .ofTypes("net.sqlcipher.database.SQLiteDatabase")
+    .names("changePassword", "openDatabase", "openOrCreateDatabase", "create")
+    .withAnyParameters()
+    .build();
+
+  private static final MethodMatchers REALM_CONFIGURATION_BUILDER_ENCRYPTION_KEY = MethodMatchers.create()
+    .ofTypes("io.realm.RealmConfiguration$Builder")
+    .names("encryptionKey")
+    .addParametersMatcher("byte[]")
+    .build();
+
+  private static final MethodMatchers STRING_TO_CHAR_GET_BYTES = MethodMatchers.create()
+    .ofTypes(JAVA_LANG_STRING)
+    .names("toCharArray", "getBytes")
+    .withAnyParameters()
+    .build();
+
+  @Override
+  protected MethodMatchers getMethodInvocationMatchers() {
+    return MethodMatchers.or(
+      SQLITE_DATABASE_CONSTRUCTOR,
+      SQLITE_DATABASE_METHODS,
+      REALM_CONFIGURATION_BUILDER_ENCRYPTION_KEY);
+  }
+
+  @Override
+  protected void onMethodInvocationFound(MethodInvocationTree mit) {
+    Arguments arguments = mit.arguments();
+    String messageArg = REALM_CONFIGURATION_BUILDER_ENCRYPTION_KEY.matches(mit) ? "encryptionKey" : "password";
+    ExpressionTree passwordArg = arguments.size() == 1 ? arguments.get(0) : arguments.get(1);
+    reportIssueIfHardCoded(passwordArg, messageArg);
+  }
+
+  @Override
+  protected void onConstructorFound(NewClassTree newClassTree) {
+    reportIssueIfHardCoded(newClassTree.arguments().get(1), "password");
+  }
+
+  private void reportIssueIfHardCoded(ExpressionTree expressionTree, String messageArg) {
+    ExpressionTree stringExpression = expressionTree;
+    if (!expressionTree.symbolType().is(JAVA_LANG_STRING)) {
+      // byte[] or char[]
+      stringExpression = tryGetOriginStringFromByteOrCharArray(expressionTree);
+      if (stringExpression == null) {
+        return;
+      }
+    }
+    ExpressionsHelper.ValueResolution<String> constantValueAsString = ExpressionsHelper.getConstantValueAsString(stringExpression);
+    if (constantValueAsString.value() != null) {
+      reportIssue(expressionTree, "The \"" + messageArg + "\" parameter should not be hardcoded.", constantValueAsString.valuePath(), null);
+    }
+  }
+
+  @CheckForNull
+  private static ExpressionTree tryGetOriginStringFromByteOrCharArray(ExpressionTree givenExpression) {
+    ExpressionTree expression = givenExpression;
+    if (expression.is(Tree.Kind.IDENTIFIER)) {
+      IdentifierTree identifier = (IdentifierTree) expression;
+      final ExpressionTree singleWriteUsage = getSingleWriteUsage(identifier.symbol());
+      if (singleWriteUsage != null && singleWriteUsage.is(Tree.Kind.METHOD_INVOCATION)) {
```

Comment:
```suggestion
      if (singleWriteUsage != null) {
```
The second part of the binary is redundant: it will be checked at line 112.